### ISSUE TITLE
GitHub Integration Proposal

### ISSUE THREAD

**Issue by berndca** [librecores/librecores-web#120]
I would like to propose an architecture for the integration of GitHub with the LibreCores website.

The main point is that the Symfony web app and anything running on the web server does not have any part in extracting the required data from the GitHub API. An external service like AWS Lambda is responsible for extracting the data from the GitHub API and storing it in the LibreCores database.

The LibreCores web server and the database are running on different EC2 instances!

## Proposed Flow

An authenticated user navigates to /project/new and enters the information for the new project in a form similar to the existing one. Upon submitting the form the information is transfered to an AWS API Gateway endpoint. The AWS Gateway triggers the executing of an AWS Lambda script which is then extracting the project information from the GitHub API. Finally the AWS Lambda script adds the new project to the LibreCores database by sending JSON requests to a (not yet existing) LibreCores API endpoint.

The “New Project” AWS Lambda script does also install a webhook in GitHub. Whenever a new commit is pushed to the master branch of this repo an API request to AWS API Gateway is issued and the AWS Lambda “Project Update” script triggered. This will keep the database up to date and avoids polling.

The same approach could be used to integrate BitBucket and Gitlab.

### [AWS Lambda](https://aws.amazon.com/lambda/)

- Programming language: Python
- Database ORM: [SQL Alchemy](http://docs.sqlalchemy.org/en/latest/dialects/mysql.html) 

There will almost certainly no charge for this service because of the very generous free tier available to all AWS customers.

### [AWS API Gateway](https://aws.amazon.com/api-gateway/)

There will be some Javascript involved on LibreCores.org/project/new and for the webhook. AWS provides an js SDK which helps somewhat.

The expected cost is below USD 1 per month. For their Frankfurt region they charge USD 3.70 per 1 million requests and USD 0.09 per GB transferred out.

## REST API for LibreCores.org

There is currently no REST API on LibreCores.org. This is something which would be required. We would also have to update the database schema to support all the data we might want to use.

**Comment by wallento** [comment]
Hi @berndca,

this sounds very interesting, despite I don't fully understand it without reading the details of the other AWS stuff :)

@imphil also mentioned he is working on abstracting the source repository capabilities to encapsulate API calls or a plain checkout. That would also extend to other properties and badges, I think.

I will need to educate myself about the pros and cons of both, but that is a good topic for our next call.

Cheers,
Stefan

**Comment by eliaskousk** [comment]
I really like the proposal of @berndca. On a quick look of it I think it seems solid and very useful for our future needs regarding the scalability of LibreCores. I've never used AWS Lambda but it seems good and easy to use. The only drawback I see is that right now we don't actually need such a nice solution and something simpler could work as well for our immediate needs until we gain more traction. If however the implementation doesn't need a lot of time to make it I say we go for it right now.

For the LibreCores REST API I propose that we use the [API Platform](https://api-platform.com/) solution (instead of the other well-known [FOSRestBundle](https://github.com/FriendsOfSymfony/FOSRestBundle) which I think is a bit outdated and harder to use. I've played with it recently and it is very easy to use on both the user and the developer. You can find a [demo here](https://demo.api-platform.com).

It can be used both as a [standalone bundle](https://api-platform.com/docs/core/) that provides an easy way to export our entities as a REST API (with modern JSON-LD and / or Hydra formats) and as a [distribution / framework on-top of Symfony](https://api-platform.com/docs/distribution/) to build an API endpoint and only that. The later is not strictly needed for us since we already have our app and thus we can start using only the bundle and add as much other functionality as we want to come close to the full framework's features.

One of the features I really like and I suggest we use is the swagger UI to visualize and interact with our api without writing a single line of code. It is also very useful as documentation. The [demo](https://demo.api-platform.com) linked above is actually the swagger UI interacting with the underlying REST API.

If you like this solution I can start a new REST API issue to start development and discuss more details on the above.

**Comment by berndca** [comment]
The programming part is rather straightforward. However the configuration of the AWS settings including IAM etc. will also take some effort. The upside is one these settings are working it becomes very easy to implement changes or clone and change the script to support another provider like Bitbucket.

We should probably review the database schemas for the models containing repository (and user/org - to put their picture in their profile #65) data to address #80.

API Platform looks great! This would be extremely helpful for features like #41 #103, #40 ...

**Comment by berndca** [comment]
@oleg-nenashev I have followed up on your concern that we might exceed GitHub's API quota when polling their API. I have created an AWS Lambda script to poll https://api.github.com/repos/ for the 43 repos currently registered on LibreCores plus another 400+ hardware related repos every 15 minutes. I have not encountered any problems with GitHub. This is not much of a surprise since I'm only making less than 1800 requests per hour when the quota is 5000. Over 48 hours during last weekend there were less than 30 change events (mostly push to master), i.e. only one in six updates saw any changes. 

This leads me to believe that a polling based API scraper for librecores is feasible and a very good match for projects which are not extremely active. The 15 minutes interval could be shortened is necessary without running into issues with the rate limit. For active projects we might consider using Webhooks to propagate any changes faster.

@wallento rather than going over the AWS Lambda features etc. you might find it more informative to browse some of the projects using this technology. GitHub has a very good [Showcase for Serverless Architecture](https://github.com/showcases/serverless-architecture) which lists quite a number of very interesting projects.

**Comment by oleg-nenashev** [comment]
> This leads me to believe that a polling based API scraper for librecores is feasible and a very good match for projects which are not extremely active. The 15 minutes interval could be shortened is necessary without running into issues with the rate limit. For active projects we might consider using Webhooks to propagate any changes faster.

For now I'm perfectly fine with the YAGNI approach. If we are far from hitting the limit, no need to overcomplicate the things

**Comment by berndca** [comment]
I agree. Polling is very easy to implement. We can always reduce the polling intervals if needed.
